Source/DafnyDriver/CliCompilation.cs

```diff
@@ -0,0 +1,228 @@
+              break;
+            case ConditionGeneration.Outcome.OutOfMemory:
+              Interlocked.Increment(ref statSum.OutOfMemoryCount);
+              break;
+            case ConditionGeneration.Outcome.OutOfResource:
+              Interlocked.Increment(ref statSum.OutOfResourceCount);
+              break;
+            case ConditionGeneration.Outcome.Inconclusive:
+              Interlocked.Increment(ref statSum.InconclusiveCount);
+              break;
+            default:
+              throw new ArgumentOutOfRangeException();
+          }
+
+          if (canVerifyResult.CompletedParts.Count == canVerifyResult.Tasks.Count) {
+            canVerifyResult.Finished.TrySetResult();
+          }
+        }
+      }
+    });
+
+    try {
+      var resolution = await Compilation.Resolution;
+      if (errorCount > 0) {
+        return;
+      }
+
+      var canVerifies = resolution.CanVerifies?.DistinctBy(v => v.Tok).ToList();
+
+      if (canVerifies != null) {
+        var orderedCanVerifies = canVerifies.OrderBy(v => v.Tok.pos).ToList();
+        foreach (var canVerify in orderedCanVerifies) {
+          canVerifyResults[canVerify] = new CliCanVerifyResults();
+          await Compilation.VerifyCanVerify(canVerify, false);
+        }
+
+        foreach (var canVerify in orderedCanVerifies) {
+          var results = canVerifyResults[canVerify];
+          try {
+            await results.Finished.Task;
+            foreach (var (task, completed) in results.CompletedParts.OrderBy(t => t.Item1.Implementation.Name)) {
+              foreach (var vcResult in completed.Result.VCResults) {
+                Compilation.ReportDiagnosticsInResult(options, task, vcResult, Compilation.Reporter);
+              }
+
+              ProofDependencyWarnings.WarnAboutSuspiciousDependenciesForImplementation(options,
+                resolution.ResolvedProgram!.Reporter,
+                resolution.ResolvedProgram.ProofDependencyManager,
+                new DafnyConsolePrinter.ImplementationLogEntry(task.Implementation.VerboseName,
+                  task.Implementation.tok),
+                DafnyConsolePrinter.DistillVerificationResult(completed.Result));
+            }
+          } catch (ProverException e) {
+            Interlocked.Increment(ref statSum.SolverExceptionCount);
+            Compilation.Reporter.Error(MessageSource.Verifier, ResolutionErrors.ErrorId.none, canVerify.Tok, e.Message);
+          } catch (Exception e) {
+            Interlocked.Increment(ref statSum.SolverExceptionCount);
+            Compilation.Reporter.Error(MessageSource.Verifier, ResolutionErrors.ErrorId.none, canVerify.Tok,
+              $"Internal error occurred during verification: {e.Message}\n{e.StackTrace}");
+          }
+        }
+      }
+
+      SynchronousCliCompilation.WriteTrailer(options, /* TODO ErrorWriter? */ options.OutputWriter, statSum);
+
+    } catch (TaskCanceledException) {
+    }
```

Comment:
When does this happen? Why not catch OperationCanceledException instead?

Reply:
Added a comment and changed to `OperationCanceledException`. Happens when there's a parsing error so it can not resolve. I could change the interface of `Compiler` so it uses `null` instead of `OperationCanceledException` when it can't return a Task<ResolutionResult` due to a user program error. This means inside `Compilation` there's a need for more `if (resultIDependOn == null) { return null }` since we don't coast on the monadic behavior of async/await, but it might be more readable.